controllers/infra/zone/zone_controller.go

```diff
@@ -0,0 +1,167 @@
+	)
+
+	r := NewReconciler(
+		ctx,
+		mgr.GetClient(),
+		ctrl.Log.WithName("controllers").WithName(controlledTypeName),
+		record.New(mgr.GetEventRecorderFor(controllerNameLong)),
+	)
+
+	return ctrl.NewControllerManagedBy(mgr).
+		For(controlledType).
+		WithOptions(controller.Options{
+			SkipNameValidation: SkipNameValidation,
+		}).
+		Complete(r)
+}
+
+func NewReconciler(
+	ctx context.Context,
+	client client.Client,
+	logger logr.Logger,
+	recorder record.Recorder) *Reconciler {
+
+	return &Reconciler{
+		Context:  ctx,
+		Client:   client,
+		Logger:   logger,
+		Recorder: recorder,
+	}
+}
+
+// Finalizer is the finalizer placed on Zone objects by VM Operator.
+const Finalizer = "vmoperator.vmware.com/finalizer"
+
+// Reconciler reconciles a StoragePolicyQuota object.
+type Reconciler struct {
+	client.Client
+	Context  context.Context
+	Logger   logr.Logger
+	Recorder record.Recorder
+}
+
+// +kubebuilder:rbac:groups=topology.tanzu.vmware.com,resources=zones,verbs=get;list;watch;update;patch
+// +kubebuilder:rbac:groups=topology.tanzu.vmware.com,resources=zones/status,verbs=get
+
+func (r *Reconciler) Reconcile(
+	ctx context.Context,
+	req ctrl.Request) (_ ctrl.Result, reterr error) {
+
+	ctx = pkgcfg.JoinContext(ctx, r.Context)
+	ctx = watcher.JoinContext(ctx, r.Context)
+
+	var obj topologyv1.Zone
+	if err := r.Get(ctx, req.NamespacedName, &obj); err != nil {
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	logger := ctrl.Log.WithName("Zone").WithValues("name", req.NamespacedName)
+
+	patchHelper, err := patch.NewHelper(&obj, r.Client)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+	defer func() {
+		if err := patchHelper.Patch(ctx, &obj); err != nil {
+			if reterr == nil {
+				reterr = err
+			} else {
+				reterr = fmt.Errorf("%w,%w", err, reterr)
+			}
+		}
+	}()
+
+	if !obj.DeletionTimestamp.IsZero() {
+		if err := r.ReconcileDelete(ctx, logger, &obj); err != nil {
+			return ctrl.Result{}, err
+		}
+	} else if err := r.ReconcileNormal(ctx, logger, &obj); err != nil {
+		return ctrl.Result{}, err
+	}
+
+	return ctrl.Result{}, nil
+}
+
+func (r *Reconciler) ReconcileDelete(
+	ctx context.Context,
+	logger logr.Logger,
+	obj *topologyv1.Zone) error {
+
+	// TODO(akutz) Index the Zone's spec.managedVMs.folderMoID field and
+	// check if other zones have this same folder ref upon deleting a
+	// zone.
+	if controllerutil.ContainsFinalizer(obj, Finalizer) {
+		if val := obj.Spec.ManagedVMs.FolderMoID; val != "" {
+			watcher.Remove(
+				ctx,
+				vimtypes.ManagedObjectReference{
+					Type:  "Folder",
+					Value: val,
+				})
+			controllerutil.RemoveFinalizer(obj, Finalizer)
+		}
+	}
+
+	return nil
+}
+
+func (r *Reconciler) ReconcileNormal(
+	ctx context.Context,
+	logger logr.Logger,
+	obj *topologyv1.Zone) error {
+
+	if !controllerutil.ContainsFinalizer(obj, Finalizer) {
+		if val := obj.Spec.ManagedVMs.FolderMoID; val != "" {
+			watcher.Add(
```

Comment:
No, it won't, because the old implementation has a goroutine inside of it. However, this code is changed and now returns immediately sans a goroutine.